### PR TITLE
replace receive slot with event stream

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

 Feature

**What does this PR do? Why is it needed?**

[/eth/v1/events](https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Events/eventstream)

Removes the receive slot function and replaces it with an event stream. 

- The gRPC implementation of the event stream still uses the stream slot endpoint but wraps it as a head event. this is a TODO item to fix in the following PR to potentially introduce a temporary event endpoint for gRPC.
- introduces a basic health checker to properly restart the event stream if disconnected.

this is a big change

### CODE REVIEW

File: api/client/event/eventStream.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+			close(eventsChannel)
+		}
+	}()
+	// Create a new scanner to read lines from the response body
+	scanner := bufio.NewScanner(resp.Body)
+
+	var eventType, data string // Variables to store event type and data
+
+	// Iterate over lines of the event stream
+	for scanner.Scan() {
```
Comment (james-prysm) [inline api/client/event/eventStream.go:104]: is a span enough...

File: api/client/event/event_stream.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+	// Iterate over lines of the event stream
+	for scanner.Scan() {
+		select {
+		case <-h.ctx.Done():
+			log.Info("Context canceled, stopping event stream")
+			close(eventsChannel)
+			return
+		default:
+			line := scanner.Text()
```
Comment (james-prysm) [inline api/client/event/event_stream.go:113]: using scanner does not account for everything in the [eventsource spec](https://html.spec.whatwg.org/multipage/server-sent-events.html#the-eventsource-interface)

I may need to do additional work to resolve cases of `\r`

there may be other things missing from the spec as well, not sure if it's appropriate to use an off-the-shelf library like 
https://github.com/launchdarkly/eventsource
https://github.com/r3labs/sse

File: validator/client/grpc-api/grpc_validator_client.go [hunk 3]
```diff
@@ -143,13 +147,80 @@ func (grpcValidatorClient) GetAggregatedSelections(context.Context, []iface.Beac
...
+	ctx, span := trace.StartSpan(ctx, "validator.gRPCClient.StartEventStream")
+	defer span.End()
+	if len(topics) == 0 {
+		eventsChannel <- &eventClient.Event{
+			EventType: eventClient.EventError,
+			Data:      []byte(errors.New("no topics were added").Error()),
+		}
+		return
+	}
+	// TODO: ONLY WORKS WITH HEAD TOPIC RIGHT NOW/ONLY PROVIDES THE SLOT
```
Comment (james-prysm) [inline validator/client/grpc-api/grpc_validator_client.go:163]: would work correctly if we had a temporary gRPC implementation

File: api/client/beacon/health.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	"sync"
+)
+
+type NodeHealth struct {
+	isHealthy     bool
+	HealthCh      chan bool
+	isHealthyLock sync.RWMutex
+}
+
+func NewNodeHealthTracker() *NodeHealth {
```
Comment (rkapka) [inline api/client/beacon/health.go:13]:
```suggestion
func NewNodeHealth() *NodeHealth {
```

File: api/client/beacon/health.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+	return &NodeHealth{
+		isHealthy: true, // just default it to true
```
Comment (rkapka) [inline api/client/beacon/health.go:15]: Defaulting to `false` is safer. Otherwise immediately after creating the struct things may start running as if the node is healthy, even if it's not. Better to wait for the node to explicitly signal that it's healthy.

File: api/client/event/event_stream_test.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+			_, err := NewEventStream(context.Background(), &http.Client{}, tt.host, tt.topics)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("NewEventStream() error = %v, wantErr %v", err, tt.wantErr)
+			}
+		})
+	}
+}
+
+func TestEventHandler(t *testing.T) {
+	//logHook := logtest.NewGlobal()
```
Comment (rkapka) [inline api/client/event/event_stream_test.go:42]: comment

File: api/client/event/event_stream.go [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+					// Process the event when both eventType and data are set
+					eventsChannel <- &Event{EventType: eventType, Data: []byte(data)}
+				}
+
+				// Reset eventType and data for the next event
+				eventType, data = "", ""
+				continue
+			}
+
+			if line[0:7] == "event: " {
```
Comment (rkapka) [inline api/client/event/event_stream.go:127]: you can use `strings.HasPrefix` and `strings.CutPrefix` for both event type and data lines

File: api/client/event/event_stream.go [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+func (h *EventStream) Subscribe(eventsChannel chan<- *Event) {
+	allTopics := strings.Join(h.topics, ",")
+	log.Info("Starting listening to Beacon API events on topics: " + allTopics)
+	fullUrl := h.host + "/eth/v1/events?topics=" + allTopics
+	req, err := http.NewRequestWithContext(h.ctx, http.MethodGet, fullUrl, nil)
+	if err != nil {
+		eventsChannel <- &Event{
+			EventType: EventConnectionError,
+			Data:      []byte(errors.Wrap(err, "Failed to create HTTP request").Error()),
```
Comment (rkapka) [inline api/client/event/event_stream.go:81]:
```suggestion
			Data:      []byte(errors.Wrap(err, "failed to create HTTP request").Error()),
```

File: api/client/beacon/health.go (hunk 5 above)
Comment (james-prysm) [reply api/client/beacon/health.go:15]: I think there's a 12 second buffer, i think it's better user experience for node runners to assume it's working first one ( which is what's happening right now anyways)

Comment (rkapka) [reply api/client/beacon/health.go:15]: We can check health at the start of `runHealthCheckRoutine` to get the initial state

Comment (james-prysm) [reply api/client/beacon/health.go:15]: hmm maybe I need to add something for initializing the health tracker then. will think about it and add something if needed.

File: validator/client/beacon-api/beacon_api_validator_client.go [hunk 9]
```diff
@@ -163,17 +156,24 @@ func (c *beaconApiValidatorClient) WaitForChainStart(ctx context.Context, _ *emp
...
+		eventsChannel <- &event.Event{
+			EventType: event.EventError,
+			Data:      []byte(errors.Wrap(err, "failed to start event stream").Error()),
 		}
+		return
 	}
-	return nil
+	c.isEventStreamRunning = true
+	eventStream.Subscribe(eventsChannel)
+	c.isEventStreamRunning = false
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply validator/client/beacon-api/beacon_api_validator_client.go:172]: subscribes runs until canceled or eof, when that happens it will come out of that function and set stream running to false

File: validator/client/runner.go [hunk 10]
```diff
@@ -128,6 +147,8 @@ func run(ctx context.Context, v iface.Validator) {
 				continue
 			}
 			performRoles(slotCtx, allRoles, v, slot, &wg, span)
+		case currentKeys := <-accountsChangedChan: // should be less of a priority than next slot
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply validator/client/runner.go:150]: events should take place before nextslot right? the event may do something that changes the duties in the slot

File: validator/client/validator.go [hunk 11]
```diff
@@ -1073,8 +1050,40 @@ func (v *validator) PushProposerSettings(ctx context.Context, km keymanager.IKey
 	return nil
 }
 
-func (v *validator) StartEventStream(ctx context.Context) error {
-	return v.validatorClient.StartEventStream(ctx)
+func (v *validator) StartEventStream(ctx context.Context, topics []string, eventsChannel chan<- *eventClient.Event) {
+	log.Infof("starting event stream for topics %v...", topics)
```
Comment (rkapka) [inline validator/client/validator.go:1054]:
```suggestion
	log.Infof("Starting event stream for topics %v...", topics)
```

File: validator/client/iface/validator_client.go [hunk 12]
```diff
@@ -86,9 +87,8 @@ type ValidatorClient interface {
 	GetSyncSubcommitteeIndex(ctx context.Context, in *ethpb.SyncSubcommitteeIndexRequest) (*ethpb.SyncSubcommitteeIndexResponse, error)
 	GetSyncCommitteeContribution(ctx context.Context, in *ethpb.SyncCommitteeContributionRequest) (*ethpb.SyncCommitteeContribution, error)
 	SubmitSignedContributionAndProof(ctx context.Context, in *ethpb.SignedContributionAndProof) (*empty.Empty, error)
-	StreamSlots(ctx context.Context, in *ethpb.StreamSlotsRequest) (ethpb.BeaconNodeValidator_StreamSlotsClient, error)
 	SubmitValidatorRegistrations(ctx context.Context, in *ethpb.SignedValidatorRegistrationsV1) (*empty.Empty, error)
-	StartEventStream(ctx context.Context) error
+	StartEventStream(ctx context.Context, topics []string, eventsChannel chan<- *event.Event)
 	EventStreamIsRunning() bool
```
Comment (rkapka) [inline validator/client/iface/validator_client.go:92]: This can be removed, it's not used anywhere. Same for the `isEventStreamRunning` field.

File: validator/client/validator.go [hunk 13]
```diff
@@ -1073,8 +1050,40 @@ func (v *validator) PushProposerSettings(ctx context.Context, km keymanager.IKey
...
+	v.validatorClient.StartEventStream(ctx, topics, eventsChannel)
+}
+
+func (v *validator) ProcessEvent(event *eventClient.Event) error {
+	if event == nil || event.Data == nil {
+		log.Warn("received empty event")
```
Comment (rkapka) [inline validator/client/validator.go:1060]:
```suggestion
		log.Warn("Received empty event")
```

File: validator/client/validator.go [hunk 14]
```diff
@@ -1073,8 +1050,40 @@ func (v *validator) PushProposerSettings(ctx context.Context, km keymanager.IKey
...
+		return nil
+	}
+	switch event.EventType {
+	case eventClient.EventConnectionError:
+		return errors.New(string(event.Data))
+	case eventClient.EventError:
+		log.Error(string(event.Data))
+		return nil // skip restart if it's not a connection error
+	case eventClient.EventHead:
+		log.Debug("received head event")
```
Comment (rkapka) [inline validator/client/validator.go:1070]:
```suggestion
		log.Debug("Received head event")
```

File: validator/client/validator.go [hunk 15]
```diff
@@ -1073,8 +1050,40 @@ func (v *validator) PushProposerSettings(ctx context.Context, km keymanager.IKey
...
+		head := &structs.HeadEvent{}
+		if err := json.Unmarshal(event.Data, head); err != nil {
+			log.Error(errors.Wrap(err, "failed to unmarshal head Event into JSON").Error())
+			return nil
+		}
+		uintSlot, err := strconv.ParseUint(head.Slot, 10, 64)
+		if err != nil {
+			log.Error(errors.Wrap(err, "failed to parse slot"))
```
Comment (rkapka) [inline validator/client/validator.go:1078]:
```suggestion
			log.WithError(err).Error("Failed to parse slot")
```

File: validator/client/validator.go [hunk 16]
```diff
@@ -1073,8 +1050,40 @@ func (v *validator) PushProposerSettings(ctx context.Context, km keymanager.IKey
...
+			return nil
+		}
+		v.setHighestSlot(primitives.Slot(uintSlot))
+	default:
+		// just keep going and log the error
+		log.Warnf("received an unknown event type of %s with contents %s", event.EventType, string(event.Data))
```
Comment (rkapka) [inline validator/client/validator.go:1084]:
```suggestion
		log.Warnf("Received an unknown event type of %s with contents %s", event.EventType, string(event.Data))
```

File: validator/client/validator.go [hunk 17]
```diff
@@ -1073,8 +1050,40 @@ func (v *validator) PushProposerSettings(ctx context.Context, km keymanager.IKey
...
+	case eventClient.EventConnectionError:
+		return errors.New(string(event.Data))
+	case eventClient.EventError:
+		log.Error(string(event.Data))
+		return nil // skip restart if it's not a connection error
+	case eventClient.EventHead:
+		log.Debug("received head event")
+		head := &structs.HeadEvent{}
+		if err := json.Unmarshal(event.Data, head); err != nil {
+			log.Error(errors.Wrap(err, "failed to unmarshal head Event into JSON").Error())
```
Comment (rkapka) [inline validator/client/validator.go:1073]:
```suggestion
			log.WithError(err).Error("Failed to unmarshal head Event into JSON")
```

File: validator/client/runner.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (rkapka) [reply validator/client/runner.go:150]: For that to happen the event would need to take place right at the start of the slot. The only event that can cause this is a reorg, but I don't think it happens at slot time 0. On the downside, if for whatever reason there is a bunch of events to process or an event takes long to process, it will delay other more important things.

File: validator/client/iface/validator_client.go (hunk 12 above)
Comment (james-prysm) [reply validator/client/iface/validator_client.go:92]: I'm confused this is used...

File: validator/client/runner.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (james-prysm) [reply validator/client/runner.go:147]: ok i implemented something that should work well for this case

File: validator/client/validator.go [hunk 18]
```diff
@@ -1073,8 +1050,33 @@ func (v *validator) PushProposerSettings(ctx context.Context, km keymanager.IKey
...
+	v.validatorClient.StartEventStream(ctx, topics, eventsChannel)
+}
+
+func (v *validator) ProcessEvent(event *eventClient.Event) {
+	if event == nil || event.Data == nil {
+		log.Warn("Received empty event")
+	}
+	switch event.EventType {
+	case eventClient.EventConnectionError, eventClient.EventError:
+		log.WithError(errors.New(string(event.Data))).Warn("event stream interrupted...")
```
Comment (rkapka) [inline validator/client/validator.go:1064]:
```suggestion
		log.WithError(errors.New(string(event.Data))).Warn("Event stream interrupted")
```
Removing the `...` because it usually suggests that the system will perform a longer running action now, which is not the case.

File: validator/client/iface/validator_client.go (hunk 12 above)
Comment (rkapka) [reply validator/client/iface/validator_client.go:92]: It was used in `runHealthCheckRoutine`, but you removed the call

File: validator/client/grpc-api/grpc_validator_client.go [hunk 19]
```diff
@@ -143,13 +147,80 @@ func (grpcValidatorClient) GetAggregatedSelections(context.Context, []iface.Beac
...
+	containsHead := false
+	for i := range topics {
+		if topics[i] == eventClient.EventHead {
+			containsHead = true
+		}
+	}
+	if !containsHead {
+		log.Warnf("gRPC only supports the head topic, topics provided are %v.. continuing stream with head topic", topics)
```
Comment (rkapka) [inline validator/client/grpc-api/grpc_validator_client.go:171]: Why would we listen to head when the caller is not interested in it? It won't help with anything. I would return an error event in such case. A warning makes sense when there is more than 1 topic including head. In that case we should inform that other topics will be ignored.

File: validator/client/grpc-api/grpc_validator_client.go [hunk 20]
```diff
@@ -143,13 +147,80 @@ func (grpcValidatorClient) GetAggregatedSelections(context.Context, []iface.Beac
...
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("Context canceled, stopping event stream")
+			close(eventsChannel)
+			c.isEventStreamRunning = false
+			return
+		default:
+			if ctx.Err() == context.Canceled {
+				log.WithError(ctx.Err()).Error("Context canceled - shutting down slots receiver")
```
Comment (rkapka) [inline validator/client/grpc-api/grpc_validator_client.go:191]: For consistency the message should be the same as when `ctx.Done`

File: validator/client/grpc-api/grpc_validator_client.go [hunk 21]
```diff
@@ -143,13 +147,80 @@ func (grpcValidatorClient) GetAggregatedSelections(context.Context, []iface.Beac
...
+	c.isEventStreamRunning = true
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("Context canceled, stopping event stream")
+			close(eventsChannel)
+			c.isEventStreamRunning = false
+			return
+		default:
+			if ctx.Err() == context.Canceled {
```
Comment (rkapka) [inline validator/client/grpc-api/grpc_validator_client.go:190]: What if there is another error in the context?

File: validator/client/grpc-api/grpc_validator_client.go [hunk 22]
```diff
@@ -143,13 +147,80 @@ func (grpcValidatorClient) GetAggregatedSelections(context.Context, []iface.Beac
...
+				log.WithError(ctx.Err()).Error("Context canceled - shutting down slots receiver")
+				c.isEventStreamRunning = false
+				return
+			}
+			res, err := stream.Recv()
+			if err != nil {
+				log.WithError(err).Error("Could not receive slots from beacon node: " + client.ErrConnectionIssue.Error())
```
Comment (rkapka) [inline validator/client/grpc-api/grpc_validator_client.go:197]: There is no point in logging here, the error will be logged in `ProcessEvent`

File: validator/client/grpc-api/grpc_validator_client.go (hunk 3 above)
Comment (rkapka) [reply validator/client/grpc-api/grpc_validator_client.go:163]: I think this is fine :-)

File: validator/client/grpc-api/grpc_validator_client.go (hunk 21 above)
Comment (james-prysm) [reply validator/client/grpc-api/grpc_validator_client.go:190]: will check

File: validator/client/runner.go [hunk 23]
```diff
@@ -128,6 +126,26 @@ func run(ctx context.Context, v iface.Validator) {
...
 			performRoles(slotCtx, allRoles, v, slot, &wg, span)
+		case isHealthyAgain := <-beaconHealthTracker.HealthUpdates():
+			if isHealthyAgain {
+				headSlot, err = initializeValidatorAndGetHeadSlot(ctx, v)
+				if err != nil {
+					log.WithError(err).Error("Failed to re initialize validator and get head slot")
+					beaconHealthTracker.UpdateNodeHealth(false)
+					continue
+				}
+				if err := v.UpdateDuties(ctx, headSlot); err != nil {
```
Comment (james-prysm) [inline validator/client/runner.go:137]: consider if this function is very costly

File: validator/client/validator.go [hunk 24]
```diff
@@ -1073,8 +1050,33 @@ func (v *validator) PushProposerSettings(ctx context.Context, km keymanager.IKey
...
+		log.Debug("Received head event")
+		head := &structs.HeadEvent{}
+		if err := json.Unmarshal(event.Data, head); err != nil {
+			log.WithError(err).Error("Failed to unmarshal head Event into JSON")
+		}
+		uintSlot, err := strconv.ParseUint(head.Slot, 10, 64)
+		if err != nil {
+			log.WithError(err).Error("Failed to parse slot")
+		}
+		v.setHighestSlot(primitives.Slot(uintSlot))
```
Comment (james-prysm) [inline validator/client/validator.go:1075]: verify if this is correct...

File: validator/client/validator.go [hunk 25]
```diff
@@ -1073,8 +1050,33 @@ func (v *validator) PushProposerSettings(ctx context.Context, km keymanager.IKey
...
+	v.validatorClient.StartEventStream(ctx, topics, eventsChannel)
+}
+
+func (v *validator) ProcessEvent(event *eventClient.Event) {
+	if event == nil || event.Data == nil {
+		log.Warn("Received empty event")
+	}
+	switch event.EventType {
+	case eventClient.EventConnectionError, eventClient.EventError:
+		log.WithError(errors.New(string(event.Data))).Warn("Event stream interrupted")
```
Comment (rkapka) [inline validator/client/validator.go:1064]:
```suggestion
		log.WithError(errors.New(string(event.Data))).Error("Event stream interrupted")
```

File: validator/client/runner.go [hunk 26]
```diff
@@ -128,6 +126,26 @@ func run(ctx context.Context, v iface.Validator) {
 				continue
 			}
 			performRoles(slotCtx, allRoles, v, slot, &wg, span)
+		case isHealthyAgain := <-beaconHealthTracker.HealthUpdates():
+			if isHealthyAgain {
+				headSlot, err = initializeValidatorAndGetHeadSlot(ctx, v)
```
[the remark this answers is not in the document]
Comment (rkapka) [reply validator/client/runner.go:131]: mid-slot, not epoch

File: validator/client/validator.go (hunk 25 above)
Comment (rkapka) [reply validator/client/validator.go:1064]: Not 100% sure which one is better, feel free to ignore this

File: validator/client/runner.go [hunk 27]
```diff
@@ -128,6 +126,26 @@ func run(ctx context.Context, v iface.Validator) {
...
+					log.WithError(err).Error("Failed to re initialize validator and get head slot")
+					beaconHealthTracker.UpdateNodeHealth(false)
+					continue
+				}
+				if err := v.UpdateDuties(ctx, headSlot); err != nil {
+					handleAssignmentError(err, headSlot)
+					beaconHealthTracker.UpdateNodeHealth(false)
+					continue
+				}
+				log.Info("event stream reconnecting...")
```
Comment (rkapka) [inline validator/client/runner.go:142]:
```suggestion
				log.Info("Event stream reconnecting...")
```

File: validator/client/runner.go [hunk 28]
```diff
@@ -288,24 +296,23 @@ func handleAssignmentError(err error, slot primitives.Slot) {
...
+				log.WithError(ctx.Err()).Error("Context cancelled")
 				return
 			}
+			newStatus := v.NodeIsHealthy(ctx)
+			tracker.UpdateNodeHealth(newStatus)
+			// in case of node returning healthy but event stream died
+			if !v.EventStreamIsRunning() {
+				log.Info("Event stream reconnecting...")
+				go v.StartEventStream(ctx, event.DefaultEventTopics, eventsChan)
+			}
```
Comment (rkapka) [inline validator/client/runner.go:315]:
```suggestion
			if newStatus && !v.EventStreamIsRunning() {
				log.Info("Event stream reconnecting...")
				go v.StartEventStream(ctx, event.DefaultEventTopics, eventsChan)
			}
```
Might be better to replace the variable name with something like `isHealthy` because `if newStatus` doesn't sound too good

Comment (james-prysm) [reply validator/client/runner.go]: cool i added it and changed the name.

**Review by rkapka — APPROVED** [review]

**Review by prestonvanloon — COMMENTED** [review]
Partial review

File: proto/prysm/v1alpha1/node.proto [hunk 29]
```diff
@@ -94,6 +101,10 @@ service Node {
     }
 }
 
+message HealthRequest {
+    uint64 syncing_status = 2;
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply proto/prysm/v1alpha1/node.proto:105]: good catch i'm not sure why i set it to 2, will change

File: validator/client/runner.go [hunk 30]
```diff
@@ -128,6 +125,24 @@ func run(ctx context.Context, v iface.Validator) {
...
+			if isHealthyAgain {
+				headSlot, err = initializeValidatorAndGetHeadSlot(ctx, v)
+				if err != nil {
+					log.WithError(err).Error("Failed to re initialize validator and get head slot")
+					beaconHealthTracker.UpdateNodeHealth(false)
+					continue
+				}
+				if err := v.UpdateDuties(ctx, headSlot); err != nil {
+					handleAssignmentError(err, headSlot)
+					beaconHealthTracker.UpdateNodeHealth(false)
```
Comment (james-prysm) [inline validator/client/runner.go:138]: re think how safe update node health to false here would be.

File: api/client/beacon/health.go [hunk 31]
```diff
@@ -0,0 +1,47 @@
+package beacon
+
+import (
+	"sync"
+)
+
+type NodeHealth struct {
```
Comment (prestonvanloon) [inline api/client/beacon/health.go:7]: Why put this under api/client? Seems to be unrelated to making API calls.

File: api/client/event/event_stream.go [hunk 32]
```diff
@@ -0,0 +1,145 @@
...
+		ctx:        ctx,
+		httpClient: httpClient,
+		host:       host,
+		topics:     topics,
+	}, nil
+}
+
+func (h *EventStream) Subscribe(eventsChannel chan<- *Event) {
+	allTopics := strings.Join(h.topics, ",")
+	log.Info("Starting listening to Beacon API events on topics: " + allTopics)
```
Comment (prestonvanloon) [inline api/client/event/event_stream.go:75]:
```suggestion
	log.WithField("topics", allTopics).Info("Listening to Beacon API events")
```

File: api/client/event/event_stream.go [hunk 33]
```diff
@@ -0,0 +1,145 @@
...
+
+	// Iterate over lines of the event stream
+	for scanner.Scan() {
+		select {
+		case <-h.ctx.Done():
+			log.Info("Context canceled, stopping event stream")
+			close(eventsChannel)
+			return
+		default:
+			line := scanner.Text() // TODO: scanner does not handle /r and does not fully adhere to https://html.spec.whatwg.org/multipage/server-sent-events.html#the-eventsource-interface
```
Comment (prestonvanloon) [inline api/client/event/event_stream.go:113]: What is the TODO item? Should you fix that before merging?

File: api/client/event/event_stream.go [hunk 34]
```diff
@@ -0,0 +1,145 @@
...
+		eventsChannel <- &Event{
+			EventType: EventConnectionError,
+			Data:      []byte(errors.Wrap(err, client.ErrConnectionIssue.Error()).Error()),
+		}
+	}
+
+	defer func() {
+		if closeErr := resp.Body.Close(); closeErr != nil {
+			log.WithError(closeErr).Error("Failed to close events response body")
+			close(eventsChannel)
```
Comment (prestonvanloon) [inline api/client/event/event_stream.go:97]: Why do you close the eventsChannel here if there was an error in closing the response body?

File: beacon-chain/rpc/prysm/v1alpha1/node/server.go [hunk 35]
```diff
@@ -45,6 +49,29 @@ type Server struct {
...
+// GetHealth checks the health of the node
+func (ns *Server) GetHealth(ctx context.Context, request *ethpb.HealthRequest) (*empty.Empty, error) {
+	_, span := trace.StartSpan(ctx, "node.GetHealth")
+	defer span.End()
+	if ns.SyncChecker.Synced() {
+		return &empty.Empty{}, nil
+	}
+	if ns.SyncChecker.Syncing() || ns.SyncChecker.Initialized() {
+		if request.SyncingStatus != 0 {
+			// override the 200 success with the provided
```
Comment (prestonvanloon) [inline beacon-chain/rpc/prysm/v1alpha1/node/server.go:61]: The provided what? incomplete sentence

File: validator/client/beacon-api/json_rest_handler.go [hunk 36]
```diff
@@ -16,13 +16,25 @@ import (
 type JsonRestHandler interface {
 	Get(ctx context.Context, endpoint string, resp interface{}) error
 	Post(ctx context.Context, endpoint string, headers map[string]string, data *bytes.Buffer, resp interface{}) error
+	GetHttpClient() *http.Client
+	GetHost() string
```
Comment (prestonvanloon) [inline validator/client/beacon-api/json_rest_handler.go:20]: Avoid using the `Get` prefix for getters. Simply call these `Host()` or `HttpClient()`

File: validator/client/grpc-api/grpc_node_client.go [hunk 37]
```diff
@@ -29,8 +30,13 @@ func (c *grpcNodeClient) ListPeers(ctx context.Context, in *empty.Empty) (*ethpb
...
 
-func (c *grpcNodeClient) IsHealthy(context.Context) bool {
-	panic("function not supported for gRPC client")
+func (c *grpcNodeClient) IsHealthy(ctx context.Context) bool {
+	_, err := c.nodeClient.GetHealth(ctx, &ethpb.HealthRequest{})
+	if err != nil {
+		log.WithError(err).Error("failed to get health of node")
+		return false
+	}
+	return true
```
Comment (prestonvanloon) [inline validator/client/grpc-api/grpc_node_client.go:39]: Why does this always return true if any response was received? Seems like this method would identify any node as "healthy" when it returned a non-200 response code.

File: validator/client/grpc-api/grpc_validator_client.go [hunk 38]
```diff
@@ -147,13 +151,97 @@ func (grpcValidatorClient) GetAggregatedSyncSelections(context.Context, []iface.
...
+	ctx, span := trace.StartSpan(ctx, "validator.gRPCClient.StartEventStream")
+	defer span.End()
+	if len(topics) == 0 {
+		eventsChannel <- &eventClient.Event{
+			EventType: eventClient.EventError,
+			Data:      []byte(errors.New("no topics were added").Error()),
+		}
+		return
+	}
+	// TODO: ONLY WORKS WITH HEAD TOPIC RIGHT NOW/ONLY PROVIDES THE SLOT
```
Comment (prestonvanloon) [inline validator/client/grpc-api/grpc_validator_client.go:167]: Why are you yelling? (all caps) :rofl:

File: validator/client/validator.go [hunk 39]
```diff
@@ -1092,8 +1069,35 @@ func (v *validator) PushProposerSettings(ctx context.Context, km keymanager.IKey
...
+			log.WithError(err).Error("Failed to unmarshal head Event into JSON")
+		}
+		uintSlot, err := strconv.ParseUint(head.Slot, 10, 64)
+		if err != nil {
+			log.WithError(err).Error("Failed to parse slot")
+		}
+		v.setHighestSlot(primitives.Slot(uintSlot))
+	default:
+		// just keep going and log the error
+		log.Warnf("Received an unknown event type of %s with contents %s", event.EventType, string(event.Data))
```
Comment (prestonvanloon) [inline validator/client/validator.go:1099]:
```suggestion
		log.WithField("type", event.EventType).WithField("data", string(event.Data)).Warn("Received an unknown event")
```

Prefer to use fields instead of templating.

File: validator/client/grpc-api/grpc_node_client.go (hunk 37 above)
Comment (james-prysm) [reply validator/client/grpc-api/grpc_node_client.go:39]: that's all this endpoint does

 // Retrieve the current health of the node.
    rpc GetHealth(HealthRequest) returns (google.protobuf.Empty) {
        option (google.api.http) = {
            get: "/eth/v1alpha1/node/health"
        };
    }

File: validator/client/grpc-api/grpc_validator_client.go (hunk 38 above)
Comment (james-prysm) [reply validator/client/grpc-api/grpc_validator_client.go:167]: big reminder that this is ducktaped together lol it's still using the old stream slots

File: api/client/event/event_stream.go (hunk 34 above)
Comment (james-prysm) [reply api/client/event/event_stream.go:97]: good question, i thought we need to clean up in this case and close the channel but can be removed if we should send to the events channel again in this case.

File: beacon-chain/rpc/prysm/v1alpha1/node/server.go (hunk 35 above)
Comment (james-prysm) [reply beacon-chain/rpc/prysm/v1alpha1/node/server.go:61]: provided request status

File: api/client/beacon/health.go (hunk 31 above)
Comment (james-prysm) [reply api/client/beacon/health.go:7]: the idea was from the client perspective this Node health object keeps track of the server ( the beacon node) so  in this case the validator uses the package api/client to track the beacon node's health. happy to move it if it doesn't make sense.

File: validator/client/grpc-api/grpc_node_client.go (hunk 37 above)
Comment (james-prysm) [reply validator/client/grpc-api/grpc_node_client.go:39]: on grpc so it should be fine.

File: validator/client/runner.go (hunk 30 above)
Comment (james-prysm) [reply validator/client/runner.go]: removed it

**Review by rkapka — DISMISSED** [review]

File: validator/client/validator.go [hunk 40]
```diff
@@ -1104,6 +1108,14 @@ func (v *validator) NodeIsHealthy(ctx context.Context) bool {
 	return v.nodeClient.IsHealthy(ctx)
 }
 
+func (v *validator) NodeHealthTracker(ctx context.Context) *beacon.NodeHealth {
+	if v.beaconNodeHealth == nil {
+		sts := v.NodeIsHealthy(ctx)
+		v.beaconNodeHealth = beacon.NewNodeHealth(sts)
```
Comment (james-prysm) [inline validator/client/validator.go:1114]: is this safe?

File: validator/client/beacon-api/beacon_api_validator_client.go [hunk 41]
```diff
@@ -198,17 +191,24 @@ func (c *beaconApiValidatorClient) WaitForChainStart(ctx context.Context, _ *emp
 	return c.waitForChainStart(ctx)
 }
 
-func (c *beaconApiValidatorClient) StartEventStream(ctx context.Context) error {
-	if c.eventHandler != nil {
-		if err := c.eventHandler.get(ctx, []string{"head"}); err != nil {
-			return errors.Wrapf(err, "could not invoke event handler")
+func (c *beaconApiValidatorClient) StartEventStream(ctx context.Context, topics []string, eventsChannel chan<- *event.Event) {
+	ctx, span := trace.StartSpan(ctx, "validator.beaconApiClient.StartEventStream")
+	defer span.End()
```
Comment (prestonvanloon) [inline validator/client/beacon-api/beacon_api_validator_client.go:196]: If this method is meant to start a stream and block until the stream is done (forever basically) then this isn't a meaningful thing to measure

```suggestion
```

File: beacon-chain/rpc/prysm/v1alpha1/node/server.go [hunk 42]
```diff
@@ -45,6 +49,29 @@ type Server struct {
 	BeaconMonitoringPort int
 }
 
+// GetHealth checks the health of the node
+func (ns *Server) GetHealth(ctx context.Context, request *ethpb.HealthRequest) (*empty.Empty, error) {
+	_, span := trace.StartSpan(ctx, "node.GetHealth")
```
Comment (prestonvanloon) [inline beacon-chain/rpc/prysm/v1alpha1/node/server.go:54]:
```suggestion
	ctx, span := trace.StartSpan(ctx, "node.GetHealth")
```
If you don't overwrite the context, then the span details are not carried forward to other requests.

Comment (prestonvanloon) [reply beacon-chain/rpc/prysm/v1alpha1/node/server.go:54]: Also, should you set some deadline, perhaps the server should respond within 10 seconds or give up waiting?

**Review by prestonvanloon — COMMENTED** [review]
Please fix the TODOs in this PR, delete the TODOs, or create associated issues to track them.

File: beacon-chain/rpc/prysm/v1alpha1/node/server.go (hunk 42 above)
Comment (james-prysm) [reply beacon-chain/rpc/prysm/v1alpha1/node/server.go]: added a deadline in the function if that's what you're suggesting, i don't think other similar functions have deadlines though, it might default to something for grpc. more tests were added though.

**Review by prestonvanloon — APPROVED** [review]
